File: c++/402. Remove K Digits.cpp
<<<<<<< HEAD
class Solution {
public:
    string removeKdigits(string num, int k) {
        string s;
        for (int i = 0; i < num.size(); i++) {
            char ch = num[i];
            while (k!=0 && !s.empty() && s.back()>ch) {
                s.pop_back();
                k--;
            }
            if (s.empty() && ch == '0')
                continue;
            s.push_back(ch);
        }

        while (!s.empty() && k!= 0) {
            s.pop_back();
            k--;
        }

        return s.empty() ? "0" : s;
    }
};

// 构造一个递增的序列，同时开头不能为0
=======
/*
	1. 如果第二位是0，则删除第一位数字
	2. 需找第一个下降的数字删除
*/
class Solution {
public:
	string removeKdigits(string num, int k) {
		while (1)
		{
			int n = num.size();
			if (n <= k || n == 0) return "0";
			if (k-- == 0) return num;
			if (num.at(1) == '0')
			{
				size_t pos = num.find_first_not_of('0', 1);
				if (pos != string::npos)
					num = num.substr(pos);
				else
					return "0";
			}
			else
			{
				int idx = 0;
				while (idx < n - 1 && num[idx] <= num[idx + 1]) idx++;
				num.erase(num.begin() + idx);
			}
		}

		return num;
	}
};
>>>>>>> 837bde34d06bb470ffcea088785f591f742870d7

File: python/402. Remove K Digits.py
class Solution(object):
    def removeKdigits(self, num, k):
        """
        :type num: str
        :type k: int
        :rtype: str
        """
        st = []
        for ch in num:
            while k and st and st[-1] > ch:
                st.pop()
                k -= 1
            st.append(ch)
        
        st = st if k<=0 else st[:-k]
        return ''.join(st).lstrip('0') or '0'
        
# 思路：尽量删除逆序的数字

if __name__ == '__main__':
    num = "1432219"
    k = 3
    su = Solution()
    print su.removeKdigits(num, k)        